firefly/kernel/intel64/cpu/cpu.cpp

```diff
@@ -3,22 +3,70 @@
 #include "firefly/intel64/cpu/ap/ap.hpp"
 #include "firefly/intel64/cpu/apic/apic.hpp"
 #include "firefly/intel64/int/interrupt.hpp"
+#include "firefly/memory-manager/allocator.hpp"
+#include "firefly/memory-manager/secondary/heap.hpp"
+#include "frg/manual_box.hpp"
+#include "frg/spinlock.hpp"
+#include "frg/vector.hpp"
 
-// Note: This is purposely bare-bones and lacking a
-// proper "per cpu" structure (i.e. gsbase).
-// There is no multi-core yet, but I'd prefer to structure
-// everything in a way so that it's easy to get it up and running now rather than later.
 namespace firefly::kernel {
-CpuData temporaryCpuInstance;
+namespace {
+frg::manual_box<CpuData> bootCpuData;
+frg::vector<CpuData *, Allocator> allCpuContexts;
+frg::ticket_spinlock allCpuContextsLock;
+}  // namespace
+
+void setupCpuContext(AssemblyCpuData *context) {
+    context->selfPointer = context;
+    wrmsr(MSR::GsBase, reinterpret_cast<uint64_t>(context));
+}
+
+void setupBootCpuContext() {
+    bootCpuData.initialize();
+    setupCpuContext(bootCpuData.get());
+}
+
+void initializeBootProccessor(uint64_t stack) {
+    setupBootCpuContext();
+    initializeThisCpu(stack);
+}
+
+void initializeApplicationProcessor(uint64_t stack) {
+    // TODO: Check with @vo1d how to do this
+    auto cpuData = new (mm::heap->allocate(sizeof(CpuData))) CpuData;
+    setupCpuContext(cpuData);
+    initializeThisCpu(stack);
+}
+
+CpuData *getCpuData(size_t k) {
+    return allCpuContexts[k];
+}
+
+CpuData *getCpuData() {
```

Comment:
I wonder if it would be better to rename it to `get{Active, Current}CpuData`

Reply:
I kinda took it from managarm, maybe getLocalCpuData?

Reply:
That's fine too.
It's your PR you can call it whatever you want

Reply:
Done